File: plot/src/main/java/net/fwitz/math/plot/color/filter/ValuesFilter.java
package net.fwitz.math.plot.color.filter;

import static java.util.Objects.requireNonNull;

public class ValuesFilter<V> implements ValuesFilterFunction<V> {
    @SuppressWarnings("unchecked")
    private static final ValuesFilter IDENTITY = new ValuesFilter("Identity", (r, i, v) -> v);

    private final String name;
    private final ValuesFilterFunction<V> delegate;

    public ValuesFilter(String name, ValuesFilterFunction<V> delegate) {
        this.name = requireNonNull(name, "name");
        this.delegate = requireNonNull(delegate, "delegate");
    }

    public String name() {
        return name;
    }

    @Override
    public V[] filter(double[] reValues, double imValue, V[] rowValues) {
        return delegate.filter(reValues, imValue, rowValues);
    }

    @SuppressWarnings("unchecked")
    public static <V> ValuesFilter<V> identity() {
        return IDENTITY;
    }

    @Override
    public String toString() {
        return "ValuesFilter[name=" + name + ']';
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/renderer/filter/ValuesFilter.kt
package net.fwitz.math.plot.renderer.filter

class ValuesFilter<V>(
    val name: String,
    private val delegate: ValuesFilterFunction<V>
) : ValuesFilterFunction<V> {
    companion object {
        private val IDENTITY: ValuesFilter<*> = ValuesFilter("Identity") { xv, yv, zv: Array<Any> -> zv }

        @Suppress("UNCHECKED_CAST")
        fun <V> identity(): ValuesFilter<V> = IDENTITY as ValuesFilter<V>
    }

    override fun invoke(xValues: DoubleArray, yValue: Double, zValues: Array<V>): Array<V> {
        return delegate(xValues, yValue, zValues)
    }

    override fun toString(): String {
        return "ValuesFilter[name=$name]"
    }
}